### PR TITLE
Allow shared projects to specify a real project to proxy too for building and testing in VS Code

### PR DESCRIPTION
**Pull request by davidwengier**
VS on my laptop is.. um.. special, so I run tests from VS Code reasonably often, but sadly if the file that has the tests is in a shared project, it doesn't work. This fixes that by allowing test and build tasks to use a different project when they're run on a file within a shared project.

@dotnet/roslyn-infrastructure let me know if modifying files in `eng` isn't allowed, or will get clobbered by arcade, and I'll move the logic to `vscode-build.ps1` and `vscode-run-tests.ps1`, but those two files are the only thing using this method.

/cc @333fred @jmarolf who I think use VS Code too

### CODE REVIEW

File: src/Analyzers/CSharp/Tests/CSharpAnalyzers.UnitTests.shproj [hunk 1]
```diff
@@ -10,4 +10,11 @@
   <PropertyGroup />
   <Import Project="CSharpAnalyzers.UnitTests.projitems" Label="Shared" />
   <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.CSharp.targets" />
+
+  <PropertyGroup>
+    <!-- We can use this to specify a path to a project that consumes this shared project
+         and our VS Code tasks will use that project when asked to build or run tests for
+         a file in this project. -->
+    <ProxyToConsumingProject>..\..\..\EditorFeatures\CSharpTest\Microsoft.CodeAnalysis.CSharp.EditorFeatures.UnitTests.csproj</ProxyToConsumingProject>
```
Comment (davidwengier) [inline src/Analyzers/CSharp/Tests/CSharpAnalyzers.UnitTests.shproj:18]: Naming suggestions welcome!

**Comment by davidwengier** [comment]
> "run/debug tests in context" commands

They don't work for me, though I admit I haven't dug in to try to work out why. Some error in OmniSharp 🤷‍♂️

**Comment by davidwengier** [comment]
Wait, why do we even have tests in a shared project? It's only used in one real project 😕

Still, this could be useful for building.

**Comment by davidwengier** [comment]
> Wait, why do we even have tests in a shared project? It's only used in one real project

This was a mistake corrected by https://github.com/dotnet/roslyn/pull/60667

**Comment by JoeRobich** [comment]
@davidwengier I think part of the problem is that repo settings are for O# to open Compilers.sln which does not contain the project 'Microsoft.CodeAnalysis.CSharp.EditorFeatures.UnitTests.csproj'.

**Comment by davidwengier** [comment]
That might be true for the in-built build and test commands, but this definitely at least fixes the "Run tests in current file" task for me.

I was using my mac last night though, and wondering if there is a Xplat.slnf we could creat for all of the cross plat stuff, so at least restore and build might work on a few more things, but thats a separate issue.

File: eng/build-utils.ps1 [hunk 2]
```diff
@@ -200,11 +200,28 @@ function Get-ProjectFile([object]$fileInfo) {
...
-          # our location didn't change. We must be at the drive root, so give up
-          return $null
+        # We didn't find a project, but lets see if we're in a shared project
+        $files = Get-ChildItem *.shproj
+        if ($files) {
+          # found a shared project, lets see if someone told us what project to use for building etc.
+          $sharedProj = [xml](Get-Content -raw $files[0])
+          $proxyFile = $sharedProj.Project.PropertyGroup | ? { $_.PSobject.Properties.Name -contains "ProxyToConsumingProject" } | Select -Expand ProxyToConsumingProject
+          if ($proxyFile -ne $null) {
+            return Get-Item($proxyFile)
```
Comment (jaredpar) [inline eng/build-utils.ps1:210]: Don't you need to effectively `Path.Combine` the parent of `$files[0]` and `$proxyFile` here since it's a relative path?

Comment (davidwengier) [reply eng/build-utils.ps1:210]: I certainly don't claim to understand powershell, but no I don't think so. The `Set-Location` call on line 220 means that this `Get-Item` is done with that parent as the working directory, so as long as the path is relative to the .shrproj file this should work fine.

Comment (jaredpar) [reply eng/build-utils.ps1:210]: Ah true. This isn't a `-recursive` search and the `Set-Location` will properly root things.

**Comment by Youssef1313** [comment]
@davidwengier Anything left on this?

> > "run/debug tests in context" commands
> 
> They don't work for me, though I admit I haven't dug in to try to work out why. Some error in OmniSharp 🤷‍♂️

Yeah the "run tests in context" doesn't work for shared projects. The error is:

image

```
Failed to run test because "System.ArgumentNullException: Value cannot be null. (Parameter 'workingDirectory')
   at OmniSharp.DotNetTest.TestManager..ctor(Project project, String workingDirectory, IDotNetCliService dotNetCli, SemanticVersion dotNetCliVersion, IEventEmitter eventEmitter, ILogger logger) in D:\\a\\1\\s\\src\\OmniSharp.DotNetTest\\TestManager.cs:line 45
   at OmniSharp.DotNetTest.VSTestManager..ctor(Project project, String workingDirectory, IDotNetCliService dotNetCli, SemanticVersion dotNetCliVersion, IEventEmitter eventEmitter, ILoggerFactory loggerFactory) in D:\\a\\1\\s\\src\\OmniSharp.DotNetTest\\VSTestManager.cs:line 29
   at OmniSharp.DotNetTest.TestManager.Create(Project project, IDotNetCliService dotNetCli, IEventEmitter eventEmitter, ILoggerFactory loggerFactory) in D:\\a\\1\\s\\src\\OmniSharp.DotNetTest\\TestManager.cs:line 70
   at OmniSharp.DotNetTest.Services.RunTestsInContextService.Handle(RunTestsInContextRequest request) in D:\\a\\1\\s\\src\\OmniSharp.DotNetTest\\Services\\RunTestsInContextService.cs:line 37
   at OmniSharp.Endpoint.EndpointHandler`2.GetFirstNotEmptyResponseFromHandlers(ExportHandler`2[] handlers, TRequest request) in D:\\a\\1\\s\\src\\OmniSharp.Host\\Endpoint\\EndpointHandler.cs:line 0
   at OmniSharp.Endpoint.EndpointHandler`2.HandleRequestForLanguage(String language, TRequest request, RequestPacket packet) in D:\\a\\1\\s\\src\\OmniSharp.Host\\Endpoint\\EndpointHandler.cs:line 234
   at OmniSharp.Endpoint.EndpointHandler`2.Process(RequestPacket packet, LanguageModel model, JToken requestObject) in D:\\a\\1\\s\\src\\OmniSharp.Host\\Endpoint\\EndpointHandler.cs:line 143
   at OmniSharp.Stdio.Host.HandleRequest(String json, ILogger logger) in D:\\a\\1\\s\\src\\OmniSharp.Stdio\\Host.cs:line 258".
```

**Comment by davidwengier** [comment]
> @davidwengier Anything left on this?

Just those ellusive green ticks :)

File: src/Analyzers/CSharp/Tests/CSharpAnalyzers.UnitTests.shproj (hunk 1 above)
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/CSharpAnalyzers.UnitTests.shproj:18]: I'm currently facing a unit test failure that happens only when test is run from CodeStyle, but not from EditorFeatures. Consider switching to CodeStyle for now?

Comment (Youssef1313) [reply src/Analyzers/CSharp/Tests/CSharpAnalyzers.UnitTests.shproj:18]: Ignore the comment. The bug causing failure in one layer but not the other could have happened the other way around.

**Pull request closed without merging**